pkg/filter/headertojson/headertojson.go

```diff
@@ -97,24 +103,82 @@ func (h *HeaderToJSON) Status() interface{} {
 	return nil
 }
 
-func (h *HeaderToJSON) encodeJSON(input map[string]interface{}) ([]byte, error) {
-	return json.Marshal(input)
+// Handle handle HTTPContext
+func (h *HeaderToJSON) Handle(ctx context.HTTPContext) string {
+	result := h.handle(ctx)
+	return ctx.CallNextHandler(result)
 }
 
-func (h *HeaderToJSON) decodeJSON(ctx context.HTTPContext) (map[string]interface{}, error) {
+func decodeMapJSON(body []byte) (map[string]interface{}, error) {
 	res := make(map[string]interface{})
-	decoder := json.NewDecoder(ctx.Request().Body())
-	err := decoder.Decode(&res)
+	err := json.Unmarshal(body, &res)
 	if err != nil && err != io.EOF {
 		return nil, err
 	}
 	return res, nil
 }
 
-// Handle handle HTTPContext
-func (h *HeaderToJSON) Handle(ctx context.HTTPContext) string {
-	result := h.handle(ctx)
-	return ctx.CallNextHandler(result)
+func decodeArrayJSON(body []byte) ([]map[string]interface{}, error) {
+	res := []map[string]interface{}{}
+	err := json.Unmarshal(body, &res)
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+	return res, nil
+}
+
+func addHeaderToMap(bodyMap map[string]interface{}, headerMap map[string]interface{}) interface{} {
+	for k, v := range headerMap {
+		bodyMap[k] = v
+	}
+	return bodyMap
+}
+
+func addHeaderToArray(bodyArray []map[string]interface{}, headerMap map[string]interface{}) interface{} {
+	for i := range bodyArray {
+		for k, v := range headerMap {
+			bodyArray[i][k] = v
+		}
+	}
+	return bodyArray
+}
+
+func firstNonControlChar(bytes []byte) (byte, error) {
+	for _, b := range bytes {
+		switch b {
+		case '\n', '\t', '\r', ' ':
+			continue
+		case '{':
+			return '{', nil
+		case '[':
+			return '[', nil
+		default:
+			return 0, errJSONEncodeDecode
+		}
+	}
+	return 0, errJSONEncodeDecode
+}
+
+func getNewBody(reqBody []byte, headerMap map[string]interface{}) (interface{}, error) {
+	char, err := firstNonControlChar(reqBody)
+	if err != nil {
+		return nil, errJSONEncodeDecode
+	}
+	if char == '{' {
+		bodyMap, err := decodeMapJSON(reqBody)
+		if err != nil {
+			return nil, errJSONEncodeDecode
+		}
+		return addHeaderToMap(bodyMap, headerMap), nil
+
+	} else if char == '[' {
+		bodyArray, err := decodeArrayJSON(reqBody)
+		if err != nil {
+			return nil, errJSONEncodeDecode
+		}
+		return addHeaderToArray(bodyArray, headerMap), nil
+	}
+	return nil, errJSONEncodeDecode
```

Comment:
```suggestion
func firstNonBlankByte(bytes []byte) byte {
	for _, b := range bytes {
		switch b {
		case '\n', '\t', '\r', ' ':
			continue
		}
		return b
	}
	return 0
}

func getNewBody(reqBody []byte, headerMap map[string]interface{}) (interface{}, error) {
	char := firstNonBlankByte(reqBody)

	if char == '{' {
		bodyMap, err := decodeMapJSON(reqBody)
		if err != nil {
			return nil, errJSONEncodeDecode
		}
		return mergeMap(bodyMap, headerMap), nil
	}
	
	if char == '[' {
		bodyArray, err := decodeArrayJSON(reqBody)
		if err != nil {
			return nil, errJSONEncodeDecode
		}
		for i := range bodyArray {
		    bodyArray[i] = mergeMap(bodyArray[i], headerMap)
		}
		return bodyArray, nil
	}
	
	return nil, errJSONEncodeDecode
```

Reply:
updated! thanks a lot for your review! learn a lot from it!